app/gui2/src/stores/graph/graphDatabase.ts

```diff
@@ -142,15 +143,14 @@ export class GraphDb {
   ) {}
 
   private nodeIdToPatternExprIds = new ReactiveIndex(this.nodeIdToNode, (id, entry) => {
-    if (entry.pattern == null) return []
-    const exprs = new Set<AstId>()
-    entry.pattern.visitRecursive((astOrToken) => exprs.add(asAstId(astOrToken.id)))
+    const exprs: AstId[] = []
+    if (entry.pattern) entry.pattern.visitRecursiveAst((ast) => exprs.push(ast.id))
     return Array.from(exprs, (expr) => [id, expr])
   })
 
   private nodeIdToExprIds = new ReactiveIndex(this.nodeIdToNode, (id, entry) => {
-    const exprs = new Set<AstId>()
-    entry.rootSpan.visitRecursive((astOrToken) => exprs.add(asAstId(astOrToken.id)))
+    const exprs: AstId[] = []
+    entry.rootSpan.visitRecursiveAst((ast) => exprs.push(ast.id))
```

Comment:
Good question. We can rely on that due to an important invariant that the AST logic takes care to maintain.

Two `Ast` objects in a module are the same object (that is, an edit to one would be immediately seen in the other) if and only if they have the same ID. We will not encounter the same object in more than one place unless some object is a child of more than one parent. That can't happen without a parent pointer being inconsistent. We ensure that such a state is never observed:
- Local edits: `Owned` is the first line of defense, ensuring that in order to insert a node somewhere it must be removed from any previous parent. Without affine types, `Owned` can't catch all cases, but `claimChild` has an assertion to immediately catch any `Owned` misuses.
- Y.Js synchronization: Merges can cause parent pointer inconsistency. This will be checked for and repaired by each client when processing updates (the ydocserver will commit the repair eagerly; anyone else's repair will only be used if they manage to commit an edit before the ydocserver acts, but that would be fine). This also covers the case where one client manages to commit multiple, conflicting edits.